# Md. Rakibul Hasan
# 2017831005


# Importing libraries and function

In [1]:
from numpy import loadtxt
from keras.layers import Dense
from keras.layers import LeakyReLU,PReLU,ELU
from keras.layers import Dropout
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix , classification_report, precision_score, accuracy_score
import keras
from keras.models import Sequential
import seaborn as sns
import matplotlib.pyplot as plt
import csv
import numpy as np


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
dataset = pd.read_csv("/content/drive/MyDrive/ML_Datasets/nn_data.txt", header=None)
print(dataset.head())

   0    1   2   3    4     5      6   7  8
0  6  148  72  35    0  33.6  0.627  50  1
1  1   85  66  29    0  26.6  0.351  31  0
2  8  183  64   0    0  23.3  0.672  32  1
3  1   89  66  23   94  28.1  0.167  21  0
4  0  137  40  35  168  43.1  2.288  33  1


In [4]:
from sklearn import preprocessing
minmax_scale = preprocessing.MinMaxScaler().fit(dataset)
scaled_dataset = minmax_scale.transform(dataset)
scaled_dataset

array([[0.35294118, 0.74371859, 0.59016393, ..., 0.23441503, 0.48333333,
        1.        ],
       [0.05882353, 0.42713568, 0.54098361, ..., 0.11656704, 0.16666667,
        0.        ],
       [0.47058824, 0.91959799, 0.52459016, ..., 0.25362938, 0.18333333,
        1.        ],
       ...,
       [0.29411765, 0.6080402 , 0.59016393, ..., 0.07130658, 0.15      ,
        0.        ],
       [0.05882353, 0.63316583, 0.49180328, ..., 0.11571307, 0.43333333,
        1.        ],
       [0.05882353, 0.46733668, 0.57377049, ..., 0.10119556, 0.03333333,
        0.        ]])

In [5]:
dataset

0    1   2   3    4     5      6   7  8
0     6  148  72  35    0  33.6  0.627  50  1
1     1   85  66  29    0  26.6  0.351  31  0
2     8  183  64   0    0  23.3  0.672  32  1
3     1   89  66  23   94  28.1  0.167  21  0
4     0  137  40  35  168  43.1  2.288  33  1
..   ..  ...  ..  ..  ...   ...    ...  .. ..
763  10  101  76  48  180  32.9  0.171  63  0
764   2  122  70  27    0  36.8  0.340  27  0
765   5  121  72  23  112  26.2  0.245  30  0
766   1  126  60   0    0  30.1  0.349  47  1
767   1   93  70  31    0  30.4  0.315  23  0

[768 rows x 9 columns]

In [6]:
# split into input (X) and output (y) variables
scaled_X = scaled_dataset[:,0:8]
scaled_y = scaled_dataset[:,8]

In [7]:
X = dataset.iloc[:,0:8]
y = dataset.iloc[:,8]

# Importing K Fold, declaring number of folds, and declaring two array for storing precision and accuracy for each fold

In [8]:
# Defining Number of Folds for K fold validation
from sklearn.model_selection import StratifiedKFold as kfold
folds = kfold(n_splits=10)
folds

StratifiedKFold(n_splits=10, random_state=None, shuffle=False)

In [9]:
# Delcaring two array for storing precision and accuracy for each folds.
precisionScores = []
accuracyScores = []

# The Neural Network function for K fold validation

In [10]:
# The Neural Network Model for K Fold Validation
def ann_model(x_train,y_train,x_test,y_test,epochs,batch_size) :
    model = Sequential();
    model.add(Dense(8, input_shape=(8,), activation='relu'))
    model.add(Dense(6, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile( optimizer = 'adam' , loss = 'binary_crossentropy', metrics = ['accuracy'])
    model.fit(x_train,y_train,epochs=epochs,batch_size = batch_size)
    # _, acuuracy , precision = model.evaluate(x_test,y_test)
    yp = model.predict(x_test)
    y_pred = [] ;
    for element in yp:
        if element > 0.5 :
            y_pred.append(1)
        else :
            y_pred.append(0)
    print(classification_report(y_test,y_pred))
    precisionScores.append(precision_score(y_test, y_pred))
    accuracyScores.append(accuracy_score(y_test, y_pred))
    

# The Neural Network function without K Fold validation but using train test split 

In [11]:
def ann_model_without_k_fold(X,y,epochs,batch_size) :
    from sklearn.model_selection import train_test_split
    X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3)
    model = Sequential();
    model.add(Dense(8, input_shape=(8,), activation='relu'))
    model.add(Dense(6, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile( optimizer = 'adam' , loss = 'binary_crossentropy', metrics = ['accuracy'])
    model.fit(X_train,y_train,epochs=epochs,batch_size = batch_size)
    yp = model.predict(X_test);
    y_pred = [] ;
    for element in yp:
        if element > 0.5 :
            y_pred.append(1)
        else :
            y_pred.append(0)
    precisionScore = precision_score(y_test, y_pred)
    accuracyScore = accuracy_score(y_test, y_pred)
    return precisionScore,accuracyScore ;

# Implementation and results without K FOLD validation but using train test split and unscaled dataset

In [12]:
precisionScore,accuracyScore = ann_model_without_k_fold(X,y,100,20)

Epoch 1/100
27/27 [==============================] - 2s 5ms/step - loss: 14.0644 - accuracy: 0.5847
Epoch 2/100
27/27 [==============================] - 0s 3ms/step - loss: 8.3441 - accuracy: 0.4953
Epoch 3/100
27/27 [==============================] - 0s 4ms/step - loss: 3.7016 - accuracy: 0.4767
Epoch 4/100
27/27 [==============================] - 0s 7ms/step - loss: 2.2133 - accuracy: 0.5233
Epoch 5/100
27/27 [==============================] - 0s 6ms/step - loss: 1.7389 - accuracy: 0.5810
Epoch 6/100
27/27 [==============================] - 0s 4ms/step - loss: 1.5654 - accuracy: 0.5587
Epoch 7/100
27/27 [==============================] - 0s 3ms/step - loss: 1.5426 - accuracy: 0.5307
Epoch 8/100
27/27 [==============================] - 0s 4ms/step - loss: 1.4077 - accuracy: 0.5512
Epoch 9/100
27/27 [==============================] - 0s 5ms/step - loss: 1.3248 - accuracy: 0.5736
Epoch 10/100
27/27 [==============================] - 0s 3ms/step - loss: 1.2587 - accuracy: 0.6089
Epoch 11

In [13]:
print("Precision :  %.2f%% " % (precisionScore))
print("Accuracy :  %.2f%% " % (accuracyScore))

Precision :  0.51% 
Accuracy :  0.66% 


In [14]:

precisionScore,accuracyScore = ann_model_without_k_fold(scaled_X,scaled_y,100,20)



Epoch 1/100
27/27 [==============================] - 2s 6ms/step - loss: 0.6802 - accuracy: 0.6574
Epoch 2/100
27/27 [==============================] - 0s 4ms/step - loss: 0.6718 - accuracy: 0.6574
Epoch 3/100
27/27 [==============================] - 0s 3ms/step - loss: 0.6658 - accuracy: 0.6574
Epoch 4/100
27/27 [==============================] - 0s 3ms/step - loss: 0.6615 - accuracy: 0.6574
Epoch 5/100
27/27 [==============================] - 0s 4ms/step - loss: 0.6583 - accuracy: 0.6574
Epoch 6/100
27/27 [==============================] - 0s 6ms/step - loss: 0.6561 - accuracy: 0.6574
Epoch 7/100
27/27 [==============================] - 0s 5ms/step - loss: 0.6544 - accuracy: 0.6574
Epoch 8/100
27/27 [==============================] - 0s 4ms/step - loss: 0.6527 - accuracy: 0.6574
Epoch 9/100
27/27 [==============================] - 0s 3ms/step - loss: 0.6514 - accuracy: 0.6574
Epoch 10/100
27/27 [==============================] - 0s 5ms/step - loss: 0.6503 - accuracy: 0.6574
Epoch 11/

In [15]:
print("Precision :  %.2f%% " % (precisionScore))
print("Accuracy :  %.2f%% " % (accuracyScore))

Precision :  0.75% 
Accuracy :  0.77% 


In [16]:
# Implementing with K Fold validation
for train_index, test_index in folds.split(scaled_X,scaled_y):
   X_train, X_test = scaled_X[train_index], scaled_X[test_index]
   y_train, y_test = scaled_y[train_index], scaled_y[test_index]
   ann_model(X_train,y_train,X_test,y_test,100,20)

Epoch 1/100
35/35 [==============================] - 1s 2ms/step - loss: 0.6929 - accuracy: 0.4732
Epoch 2/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6853 - accuracy: 0.6570
Epoch 3/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6744 - accuracy: 0.6512
Epoch 4/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6639 - accuracy: 0.6512
Epoch 5/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6571 - accuracy: 0.6512
Epoch 6/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6534 - accuracy: 0.6512
Epoch 7/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6503 - accuracy: 0.6512
Epoch 8/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6480 - accuracy: 0.6512
Epoch 9/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6463 - accuracy: 0.6512
Epoch 10/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6443 - accuracy: 0.6512
Epoch 11/

3/3 [==============================] - 0s 3ms/step
              precision    recall  f1-score   support

         0.0       0.81      0.76      0.78        50
         1.0       0.60      0.67      0.63        27

    accuracy                           0.73        77
   macro avg       0.70      0.71      0.71        77
weighted avg       0.74      0.73      0.73        77

Epoch 1/100
35/35 [==============================] - 1s 5ms/step - loss: 0.6617 - accuracy: 0.6512
Epoch 2/100
35/35 [==============================] - 0s 6ms/step - loss: 0.6561 - accuracy: 0.6512
Epoch 3/100
35/35 [==============================] - 0s 6ms/step - loss: 0.6527 - accuracy: 0.6512
Epoch 4/100
35/35 [==============================] - 0s 7ms/step - loss: 0.6499 - accuracy: 0.6512
Epoch 5/100
35/35 [==============================] - 0s 4ms/step - loss: 0.6473 - accuracy: 0.6512
Epoch 6/100
35/35 [==============================] - 0s 3ms/step - loss: 0.6445 - accuracy: 0.6512
Epoch 7/100
35/35 [=========

3/3 [==============================] - 0s 11ms/step
              precision    recall  f1-score   support

         0.0       0.73      0.88      0.80        50
         1.0       0.65      0.41      0.50        27

    accuracy                           0.71        77
   macro avg       0.69      0.64      0.65        77
weighted avg       0.70      0.71      0.69        77

Epoch 1/100
35/35 [==============================] - 2s 2ms/step - loss: 0.6691 - accuracy: 0.6512
Epoch 2/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6653 - accuracy: 0.6512
Epoch 3/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6629 - accuracy: 0.6512
Epoch 4/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6595 - accuracy: 0.6512
Epoch 5/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6572 - accuracy: 0.6512
Epoch 6/100
35/35 [==============================] - 0s 2ms/step - loss: 0.6540 - accuracy: 0.6512
Epoch 7/100
35/35 [========

In [17]:
precisionScores 

[0.6153846153846154,
 0.8235294117647058,
 0.6956521739130435,
 0.6,
 0.6470588235294118,
 0.6818181818181818,
 0.7727272727272727,
 0.7391304347826086,
 0.631578947368421,
 0.8421052631578947]

In [18]:
accuracyScores

[0.7272727272727273,
 0.7922077922077922,
 0.7662337662337663,
 0.7272727272727273,
 0.7142857142857143,
 0.7532467532467533,
 0.8051948051948052,
 0.7922077922077922,
 0.7236842105263158,
 0.8289473684210527]

In [19]:
print("Avg Precision :  %.2f%% (+/- %.2f%%)" % (np.mean(precisionScores), np.std(precisionScores)))
print("Avg Accuracy : %.2f%% (+/- %.2f%%)" % (np.mean(accuracyScores), np.std(accuracyScores)))

Avg Precision :  0.70% (+/- 0.08%)
Avg Accuracy : 0.76% (+/- 0.04%)


# Summary and Result




*   For Unscaled Dataset >>>  **Precision : 0.51 , Accuracy : 0.66** ,
*   For Scaled Dataset Without K FOLD Validation >>> **Precision : 0.75, Accuracy : 0.77,**
*   For Scaled Dataset With K FOLD Validation >>> **Precision : 0.70 , Accuracy : 0.76**



